File: packages/core-py/src/astropods_adapter_core/auth/alb.py
from __future__ import annotations

import threading
import time
from typing import Any, Callable, Optional

from .types import Principal

IDENTITY_HEADER = "x-amzn-oidc-identity"
DATA_HEADER = "x-amzn-oidc-data"
ACCESS_TOKEN_HEADER = "x-amzn-oidc-accesstoken"


def region_from_signer_arn(arn: Optional[str]) -> Optional[str]:
    if not arn:
        return None
    parts = arn.split(":")
    return parts[3] if len(parts) > 3 and parts[3] else None


class AlbIdentityVerifier:
    """Verifies the signed claims JWT the ALB injects on authenticated requests.

    ALB puts ``exp`` in the JWT *header*, not the payload, and the payload holds
    only the IdP's user claims. Expiry is therefore checked against the header.
    """

    def __init__(
        self,
        region: Optional[str] = None,
        fetch_key: Optional[Callable[[str, str], str]] = None,
    ) -> None:
        self._region = region
        self._fetch_key = fetch_key or _fetch_alb_public_key
        self._keys: dict[str, Any] = {}
        self._lock = threading.Lock()

    def verify(self, data: str) -> Principal:
        import jwt

        header = jwt.get_unverified_header(data)

        if header.get("alg") != "ES256":
            raise ValueError(f"alb identity: unexpected alg {header.get('alg')}")

        kid = header.get("kid")
        if not kid:
            raise ValueError("alb identity: header missing kid")

        exp = header.get("exp")
        if isinstance(exp, (int, float)) and exp <= time.time():
            raise ValueError("alb identity: token expired")

        region = self._region or region_from_signer_arn(header.get("signer"))
        if not region:
            raise ValueError("alb identity: cannot determine region for key lookup")

        claims = jwt.decode(
            data,
            self._key_for(kid, region),
            algorithms=["ES256"],
            options={"verify_exp": False, "verify_aud": False},
        )

        user_id = claims.get("sub")
        if not isinstance(user_id, str) or not user_id:
            raise ValueError("alb identity: payload missing sub")

        email = claims.get("email")
        name = claims.get("name")
        return Principal(
            user_id=user_id,
            source="alb",
            email=email if isinstance(email, str) else None,
            name=name if isinstance(name, str) else None,
            claims=claims,
        )

    def _key_for(self, kid: str, region: str) -> Any:
        with self._lock:
            cached = self._keys.get(kid)
        if cached is not None:
            return cached

        from cryptography.hazmat.primitives.serialization import load_pem_public_key

        key = load_pem_public_key(self._fetch_key(kid, region).encode())
        with self._lock:
            self._keys[kid] = key
        return key


def _fetch_alb_public_key(kid: str, region: str) -> str:
    import urllib.parse

    import httpx

    url = (
        f"https://public-keys.auth.elb.{region}.amazonaws.com/"
        f"{urllib.parse.quote(kid, safe='')}"
    )
    res = httpx.get(url, timeout=5.0)
    if res.status_code != 200:
        raise ValueError(f"alb identity: key fetch returned {res.status_code}")
    return res.text

File: packages/core/src/auth/alb.ts
import { importSPKI, jwtVerify, decodeProtectedHeader } from "jose";

import type { Principal } from "./types.js";

export const IDENTITY_HEADER = "x-amzn-oidc-identity";
export const DATA_HEADER = "x-amzn-oidc-data";
export const ACCESS_TOKEN_HEADER = "x-amzn-oidc-accesstoken";

export interface AlbIdentityVerifierOptions {
  region?: string;
  /** Overridable for tests. Returns a PEM-encoded SPKI public key. */
  fetchKey?: (kid: string, region: string) => Promise<string>;
}

/**
 * Verifies the signed claims JWT the ALB injects on authenticated requests.
 *
 * ALB puts `exp` in the JWT *header*, not the payload, and the payload holds
 * only the IdP's user claims. Expiry is therefore checked against the header.
 */
export class AlbIdentityVerifier {
  private readonly keys = new Map<string, CryptoKey>();
  private readonly region?: string;
  private readonly fetchKey: (kid: string, region: string) => Promise<string>;

  constructor(options: AlbIdentityVerifierOptions = {}) {
    this.region = options.region;
    this.fetchKey = options.fetchKey ?? fetchAlbPublicKey;
  }

  async verify(data: string): Promise<Principal> {
    const header = decodeProtectedHeader(data) as {
      kid?: string;
      signer?: string;
      exp?: number;
      alg?: string;
    };

    if (header.alg !== "ES256") {
      throw new Error(`alb identity: unexpected alg ${header.alg}`);
    }
    if (!header.kid) {
      throw new Error("alb identity: header missing kid");
    }
    if (typeof header.exp === "number" && header.exp * 1000 <= Date.now()) {
      throw new Error("alb identity: token expired");
    }

    const region = this.region ?? regionFromSignerArn(header.signer);
    if (!region) {
      throw new Error("alb identity: cannot determine region for key lookup");
    }

    const key = await this.keyFor(header.kid, region);
    const { payload } = await jwtVerify(data, key, { algorithms: ["ES256"] });

    const userId = typeof payload.sub === "string" ? payload.sub : "";
    if (!userId) {
      throw new Error("alb identity: payload missing sub");
    }

    return {
      userId,
      email: typeof payload.email === "string" ? payload.email : undefined,
      name: typeof payload.name === "string" ? payload.name : undefined,
      source: "alb",
      claims: payload as Record<string, unknown>,
    };
  }

  private async keyFor(kid: string, region: string): Promise<CryptoKey> {
    const cached = this.keys.get(kid);
    if (cached) return cached;
    const pem = await this.fetchKey(kid, region);
    const key = await importSPKI(pem, "ES256");
    this.keys.set(kid, key);
    return key;
  }
}

export function regionFromSignerArn(arn: string | undefined): string | undefined {
  if (!arn) return undefined;
  const parts = arn.split(":");
  return parts.length > 3 && parts[3] ? parts[3] : undefined;
}

async function fetchAlbPublicKey(kid: string, region: string): Promise<string> {
  const url = `https://public-keys.auth.elb.${region}.amazonaws.com/${encodeURIComponent(kid)}`;
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`alb identity: key fetch returned ${res.status}`);
  }
  return res.text();
}
